rfcs/0004-checkpoints.md

```diff
@@ -0,0 +1,315 @@
+
+$ slatedb delete-checkpoint --help
+Delete an existing checkpoint
+
+Usage: slatedb --path <PATH> delete-checkpoint --id <ID>
+
+Options:
+  -i, --id <ID>  The UUID of the checkpoint (e.g. 01740ee5-6459-44af-9a45-85deb6e468e3)
+```
+
+### Writers
+
+Writers will create checkpoints to “pin” the set of SRs currently being used to serve reads. As the writer consumes manifest updates, it will re-establish its checkpoint to point to the latest manifest version. Eventually, when we support range scans, the writer will retain older checkpoints to ensure that long-lived scans can function without interference from the GC.
+
+#### Establishing Initial Checkpoint
+
+The writer first establishes it’s checkpoint when starting up, as part of incrementing the epoch:
+1. Read the latest manifest version V.
+2. Compute the writer’s epoch E
+3. Look through the checkpoints under `checkpoint` for any entries with `metadata` of type `WriterCheckpoint`, and remove any such entries.
+4. Create a new checkpoint with id set to a new v4 uuid, `manifest_id` set to the next manifest version (V + 1), and `checkpoint_expire_time_s` set to None, and `metadata` set to `WriterCheckpoint{ epoch: E }`.
+5. Set `writer_epoch` to E
+6. Write the manifest. If this fails with a CAS error, go back to 1.
+
+#### Refreshing the Checkpoint
+
+The writer will periodically re-establish its checkpoint to pin the latest SSTs and SRs. It will run the following whenever it detects that the manifest has been updated. This happens either when the manifest is polled (at `manifest_poll_interval`), or when the writer detects an update when it goes to update the manifest with it’s local view of the core db state:
+1. Read the latest manifest version V. (If the writer is applying an update from the local state then in the first iteration it can assume that it’s local version is the latest. If it’s wrong, the CAS will fail and it should reload the manifest from Object Storage))
+2. Decide whether it needs to re-establish it’s checkpoint. The writer reestablishes the checkpoint whenever the set of L0 SSTs, or Sorted Runs has changed. This includes both changes made locally by the Writer, or by the Compactor.
+    1. If so, create a new checkpoint with `manifest_id` set to V+1, `checkpoint_expire_time` set to None, and `metadata` set to `WriterCheckpoint{ epoch: E }`.
+    2. Delete the old checkpoint from `checkpoints`. When we support range scans, we’ll instead want to maintain references to the checkpoint from the range scans and delete the checkpoint when all references are dropped. (We should technically do this for point lookups too, but the likelihood of GC deleting an SST in this case is very low).
```

Comment:
The read and checkpoint/manifest maintenance happen in different tasks, so there's always a window for a race condition.

Today this looks like:
1. `get` call obtains references to the current set of SSTs/SRs so it can do its search. This includes SST S which is deleted in the latest manifest (not read by the Writer yet)
2. GC deletes S
3. `get` tries to read S and fails

Once we have checkpoints this looks like:
1. `get` call obtains references to the current set of SSTs/SRs so it can do its search. This includes SST S which is deleted in the latest manifest (not read by the Writer yet), but is in the Writer's checkpoint C.
2. GC sees S is deleted in latest manifest but doesn't delete it because the Writer holds a checkpoint
3. Writer refreshes the manifest, and establishes new checkpoint C' and deletes C
4. GC deletes S because C is removed.
5. `get` tries to read S and fails

So to be totally safe we should really have `get` take a reference to C and have the Writer delete C only when all references to it are removed. I feel like once we have https://github.com/slatedb/slatedb/issues/204 this issue becomes super unlikely for point lookups and is maybe not worth the reference tracking overhead.